lib/Sema/SemaBounds.cpp

```diff
@@ -2234,7 +2222,18 @@ namespace {
     }
 
     static bool EqualValue(ASTContext &Ctx, Expr *E1, Expr *E2, EquivExprSets *EquivExprs) {
-      Lexicographic::Result R = Lexicographic(Ctx, EquivExprs).CompareExpr(E1, E2);
+	  // It is assumed that E1 and E2 have been successfully evaluated, so that dynamic and assume bounds casts can be treated as value preserving here.
```

Comment:
I think you need to push this check into Lexicographic expression comparison.  There should be a special mode that answer the questions: if e1 and e2 evaluate to values, are the values guaranteed to be equal?  We can assume there that any evaluations of dynamic_bounds_cast in subexpressions of e1 (or 32) will succeed and any bounds checks that depend on bounds changed by *_cast operation will also succeed.